crates/runtime/src/mmap.rs

```diff
@@ -44,6 +47,74 @@ impl Mmap {
         Self::accessible_reserved(rounded_size, rounded_size)
     }
 
+    /// Creates a new `Mmap` from the specified open `file` with the `len`
+    /// specified.
+    ///
+    /// The memory is mapped in read-only mode for the entire file. If portions
+    /// of the file need to be modified then the `region` crate can be use to
+    /// alter permissions of each page.
+    ///
+    /// Note that `len` doesn't have to be page-aligned, but rather it's
+    /// expected to be the length of the `file` provided.
+    pub fn new_file(file: File, len: usize) -> Result<Self> {
+        #[cfg(unix)]
+        {
+            use std::os::unix::prelude::*;
+            let ptr = unsafe {
+                libc::mmap(
+                    ptr::null_mut(),
+                    len,
+                    libc::PROT_READ,
+                    libc::MAP_PRIVATE,
+                    file.as_raw_fd(),
+                    0,
+                )
+            };
+            if ptr as isize == -1_isize {
+                return Err(io::Error::last_os_error())
+                    .context(format!("mmap failed to allocate {:#x} bytes", len));
+            }
+
+            Ok(Self {
+                ptr: ptr as usize,
+                len,
+                file: Some(file),
+            })
+        }
+        #[cfg(windows)]
+        {
+            use std::os::windows::prelude::*;
+            use winapi::um::handleapi::*;
+            use winapi::um::memoryapi::*;
+            use winapi::um::winnt::*;
+            unsafe {
+                let mapping = CreateFileMappingW(
+                    file.as_raw_handle().cast(),
+                    ptr::null_mut(),
+                    PAGE_READONLY,
```

Comment:
I think we probably want this to be `PAGE_EXECUTE_READ` with a `FILE_MAP_COPY` below.

Reply:
Doh:

> When copy-on-write access is specified, the system and process commit charge taken is for the entire view 
> because the calling process can potentially write to every page in the view, making all pages private.

We're going to need separate mapped views for the file, one specifically for the text section that's mapped `FILE_MAP_READ`, `FILE_MAP_COPY`, and `FILE_MAP_EXECUTE`.

Reply:
Alternatively, I think we may be able to keep the single view, but change the protection level of the text section's pages to `PAGE_EXECUTE_WRITECOPY` during relocation, provided the mapping was created with `PAGE_EXECUTE_READ`?

I'd have to experiment a bit.